### PR TITLE
Improve dashboard title

### PR DESCRIPTION
**Pull request by antonymilne**
## Description

Small precursor to #185, mentioned as part of #176.

When `Dashboard.title` is specified we now include it in the page title in the dash registry. This makes no difference to anything on the screen other than the `<title>` tag and a meta tag. These are useful to get right because it means that the title of the page is much more meaningful in your browser window, when you try to bookmark a dashboard, and if you share a link of a Vizro dashboard that gets previewed e.g. on social media.

## Screenshot

image

## Checklist

- [x] I have not referenced individuals, products or companies in any commits, directly or indirectly
- [x] I have not added data or restricted code in any commits, directly or indirectly
- [x] I have updated the docstring of any public function/class/model changed
- [x] I have added tests to cover my changes (if applicable)

## Types of changes

- [ ] Docs/refactoring (non-breaking change which improves codebase)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: vizro-core/changelog.d/20231218_204302_antony.milne_improve_dashboard_title.md
Comment (antonymilne) [inline vizro-core/changelog.d/20231218_204302_antony.milne_improve_dashboard_title.md:23]: This is meant to be a "non-expert user understandable" way of saying `<title>` and is inspired by the description of Dash's `title` argument.

File: vizro-core/examples/default/app.py
Comment (antonymilne) [inline vizro-core/examples/default/app.py:548]:
```suggestion
```

File: vizro-core/src/vizro/models/_dashboard.py
Comment (antonymilne) [inline vizro-core/src/vizro/models/_dashboard.py:73]: This one line is the only functional change in the PR.

File: vizro-core/tests/unit/vizro/conftest.py
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/conftest.py:63]: I've wanted to move this to the navigation conftest for a while, and now that I've removed all its other uses from test_dashboard.py it has been set free.

File: vizro-core/tests/unit/vizro/models/test_dashboard.py
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_dashboard.py:156]: This was not a good way of doing the comparison before: actually comparing two dictionaries works fine and does not need `str`. The actual problem was the comparison of the `layout` functions, which is very tricky to get right, hence why I ignore them here.

File: vizro-core/tests/unit/vizro/models/test_dashboard.py
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/test_dashboard.py:70]: Generally we should avoid mocking much, but this is a good case where it really helps the tests become more focussed.

**Comment by antonymilne** [comment]
@AnnMarieW: I wanted to add you as a reviewer for this but for some reason can't do so - maybe some limitation of our github org settings I guess 🤔 

Are you able to add yourself as reviewer?

There's two reasons I wanted to add you anyway:
1. so you can check I haven't done anything stupid
2. I was thinking about the tests in your PR and decided our current tests weren't quite right so rewrote them

I think we should aim to get this merged ASAP before #185 so you can base your tests on the "new" style.

**Comment by AnnMarieW** [comment]
@antonymilne I might have to be invited as a collaborator or a member before I can be added as a reviewer, but I think tagging me here works too :-) 

I like the new style of the tests and will wait until this is merged before writing the tests in my PR.  Thanks for the help!

I also like the new format for the title. Are you planning on updating the docs in a different PR?

File: vizro-core/changelog.d/20231218_204302_antony.milne_improve_dashboard_title.md
Comment (huong-li-nguyen) [reply vizro-core/changelog.d/20231218_204302_antony.milne_improve_dashboard_title.md:23]: Just wondering whether we should align the naming - I think we refer to it as dashboard title while Dash refers to it as app title. Dashboard title makes more sense to me as it's an argument in the dashboard model, but under the hood we seem to overwrite the app title

This might also be just worth mentioning in the docs for people coming from Dash

File: vizro-core/examples/default/app.py
Comment (huong-li-nguyen) [inline vizro-core/examples/default/app.py:548]: You probably just left this in for demo purposes, but I would remove it if this is close to being merged. It's not optimal in terms of design if there are controls/navigation on the page, as we still have this open discussion of where this might go. 

Looking at this, I prefer the upper header container we discussed the other day even more 😄

File: vizro-core/src/vizro/models/_dashboard.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_dashboard.py:73]: I guess if people do want to overwrite it, they would overwrite the pre-build of the model here?

I was actually wondering how this works compared to the `dash.app.title`. If you set the `dash.app.title`, I noticed that it appears for a second and then it gets overwritten by anything defined in `dash.register_page`.

This might be worth mentioning in the docs if people try to overwrite it via the `dash.app.title`. This wouldn't actually work, because in the end it gets overwritten by the code here again. But I think this is also how it works in a pure Dash app?

```
app = Vizro(assets_folder="../assets").build(dashboard)
app.title = "My title"
app.run()
```

File: vizro-core/tests/unit/vizro/models/test_dashboard.py
Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/test_dashboard.py:156]: Yeees, the layout was the problem. Even better if we are fine ignoring it 👍

File: vizro-core/tests/unit/vizro/models/test_dashboard.py
Comment (huong-li-nguyen) [inline vizro-core/tests/unit/vizro/models/test_dashboard.py:109]: I really like these new tests - looks much cleaner! 🚀

**Review by huong-li-nguyen — APPROVED** [review]
Looks great to me 👍 

If we can just mention in the docs how the `dash.app.title` and `dashboard.title/page.title ` overwriting works - I think that would help. I wasn't e.g. sure how it would actually work until I tested it out and saw that the `dash.app.title` is always overwritten by the title defined in the `dash.page_registry`. I guess that's the default behaviour coming from dash?

File: vizro-core/examples/default/app.py
Comment (antonymilne) [reply vizro-core/examples/default/app.py:548]: Actually I left it here so that we would hopefully apply the same change to vizro.mckinsey.com next time we update the demo, but let me remove it if we're not yet happy about the visual design.

The reason I think it would be nice to include sooner rather than later is because every time someone shares a link to vizro.mckinsey.com the social media preview will have a slightly strange label (like just "Homepage") - it would be nicer if it instead said "Vizro demo: Homepage" or similar, to make it clear what the link being shared actually is.

If we don't work out the page layout/visual design soon then possible alternatives might be:
* we remove the title from the dashboard layout altogether from `_make_page_layout` for now
* we remove the title from the dashboard layout just by customising our demo app

So I'll revert the change here and leave it up to you to decide out what's the best solution here depending on how long the reworking of the page layout takes 🙂 It's not a huge thing but I don't think we should wait too long (say > 1 month) to get an improved shareable link to our demo app just because we're blocked on where to put the title on screen.

File: vizro-core/src/vizro/models/_dashboard.py
Comment (antonymilne) [reply vizro-core/src/vizro/models/_dashboard.py:73]: This is indeed how it works in a pure Dash app also:
1. `Dash(title="X")` corresponds to the property `dash.app.title` and sets the initial value of `<title>X</title>`
2. but this only appears very briefly when you're using Dash pages, because it gets [overwritten straight away](https://github.com/plotly/dash/blob/7527383050b6fe59a11833bbeac6f1a741436084/dash/dash.py#L2159) by  `dash.page_registry[page.id]["title"]`
3. after this, the original app `title=X` has no effect on anything

So for Vizro users where Dash pages is always used, the `title` argument basically has no effect. @AnnMarieW please do correct me if I've missed something here.

We have possibly added to the confusion here because we have our own `Dashboard.title` property, and it's not obvious to a user what effect that has on anything. I considered before whether we should do `dash.app.title = dashboard.title` in `Vizro.build()` but decided it was basically pointless because, like you say, it just gets overwritten straight away. I think it would be a very small improvement though, since that way you wouldn't see the "Dash" title at all, so let me add it in here.

File: vizro-core/changelog.d/20231218_204302_antony.milne_improve_dashboard_title.md
Comment (antonymilne) [reply vizro-core/changelog.d/20231218_204302_antony.milne_improve_dashboard_title.md:23]: Yeah, I think "dashboard title" is best here because currently the only way a user might set their title is through `Dashboard.title` since we don't advertise the `Vizro(title=...)` method anywhere.

So I will leave this alone but add something to the docs that clarifies.

File: vizro-core/examples/default/app.py
Comment (huong-li-nguyen) [reply vizro-core/examples/default/app.py:548]: **we remove the title from the dashboard layout altogether from _make_page_layout for now**
I think this would be a breaking change right? Or are we still fine saying that any visual changes are not considered breaking changes? 😄 I would be fine with this one to be honest - if people reach out, then we at least know people have been using a dashboard title and need it 😅 

**we remove the title from the dashboard layout just by customising our demo app**
As a quick-fix, this one seems safer. The demo app would be a bit out of sync for a while with the example app here, but I think it's fine as we manage it.

Comment (antonymilne) [reply vizro-core/examples/default/app.py]: Reverted in [5cef34a](https://github.com/mckinsey/vizro/pull/228/commits/5cef34afdef26e2baf0a5ee4a4302beeb61fe5ea).

File: vizro-core/src/vizro/models/_dashboard.py
Comment (antonymilne) [reply vizro-core/src/vizro/models/_dashboard.py:73]: Change made in [c00fe4e](https://github.com/mckinsey/vizro/pull/228/commits/c00fe4ed0b58ab9b13645ab4aeb8484b08dde524).

**Comment by antonymilne** [comment]
> @antonymilne I might have to be invited as a collaborator or a member before I can be added as a reviewer, but I think tagging me here works too :-)
> 
> I like the new style of the tests and will wait until this is merged before writing the tests in my PR. Thanks for the help!
> 
> I also like the new format for the title for the browser and the meta tags. Are you planning on updating the docs in a different PR?

I've just updated the docs in [7f981fb](https://github.com/mckinsey/vizro/pull/228/commits/7f981fb1c327ef11f9dfd295a61950124f184750).

When we write docs for your meta tag image and description pieces, I think it would belong in pages.md and then let's link to it from my text that mentions meta elements here in dashboard.md.

I've changed a github setting that should mean that all users can now review PRs, but I still don't seem to be able to add you here. Maybe it will work in future PRs though, let's see.

File: vizro-core/examples/default/app.py
Comment (antonymilne) [reply vizro-core/examples/default/app.py]: I wouldn't consider removing the title from the layout a breaking change since it's just something visual and no one's code will break in any way. Admittedly it's kind of a drastic one rather than just tweaking styling, but I'm completely fine with it and, like you say, if people ask where their title went then it's a good signal they want it there.

Which solution we go for depends to me on how close we are to resolving the question of where the title goes and how upsetting people find the current placement:
1. if we're really not happy with how the dashboard title appears at the moment and it feels a long way off resolving let's go for option 1
2. if we think it's ok currently but still don't want it on our demo app then let's go for 2
3. if we think the current dashboard title is already completely fine then let's just leave it as it is and add a title to the example app

Personally I'm completely fine with the current visual appearance, which is why I just went for option 3 to begin with in this PR. But you have a better idea of how upsetting people with more design sensibilities find that and where future changes to layout might take us here, so completely up to you which option to choose 🙂

File: vizro-core/examples/default/app.py
Comment (antonymilne) [inline vizro-core/examples/default/app.py:548]:
```suggestion
```

File: vizro-core/examples/default/app.py
Comment (antonymilne) [reply vizro-core/examples/default/app.py]: Cool, thank you! Sounds good to me.

I didn't actually realise how weird it looks like when there's no controls/navigation until I just looked again now. You're right that doesn't look good at all, so I take back my statement above that option 3 was ok by me. Just in case you thought I had no design sensibilities at all 😀

Comment (antonymilne) [reply vizro-core/examples/default/app.py]: P.S. hopefully the demo update would be as simple as `dash.layout["dashboard_title_outer"].hidden = True` or similar rather than needing any customisation of `_make_page_layout` with a custom `Dashboard` model.

**Pull request merged**
